compaction/compaction.cc

```diff
@@ -420,7 +420,9 @@ class formatted_sstables_list {
     bool _include_origin = true;
     std::vector<sstring> _ssts;
 public:
-    formatted_sstables_list() = default;
+    formatted_sstables_list(size_t n) {
+        _ssts.reserve(n);
+    }
```

Comment:
Also operator<< can benefit from fmt::join so to stream the strings in the vector into the output buffer.
I'm not sure that `os << boost::algorithm::join(lst._ssts, ",")` will do that, and will instead create a single massive string that could also be a source for stalls.

@tchaikov what do you think?

Reply:
I'll change it to follow std convention. as container(size_t) means pre-allocating n entries.

Reply:
> Also operator<< can benefit from fmt::join so to stream the strings in the vector into the output buffer. I'm not sure that `os << boost::algorithm::join(lst._ssts, ",")` will do that, and will instead create a single massive string that could also be a source for stalls.
> 
> @tchaikov what do you think?

note that this "massive" string amount to ~100k with 1k list, but if we can lazily assemble it into logging, that's the best.

Reply:
assembling multiple strings together with join() can produce a terrible amount of reallocations and copying too. That's indeed another potential source of stalls.

```
"    basic_sstring operator+(const basic_sstring& x) const {
        basic_sstring ret(initialized_later(), size() + x.size());
        std::copy(begin(), end(), ret.begin());
        std::copy(x.begin(), x.end(), ret.begin() + size());
        return ret;
    }
```

Reply:
Let B be avg number of bytes per string, and E be the number of entries in the list.

E - 1 = number of reallocations

The final string size will be ~(B * E) 

The total amount of copying as a result of all reallocations = (B * E) * (E - 1)

With B = 100, and E = 1000, the total amount of copying = 99900000 = ~99M

With 2020 hardware, we achieve that in less than 1ms. But it's indeed expensive.

@tchaikov I also would love to have ur input on this.

Reply:
> The problem isn't here. It's in boost::algorithm::join() interaction with sstring.
> 
> join() assumes that appending repeatedly into the string is O(log n), but it's O(n^2) with sstring, since it doesn't have a reservation and exponential capacity increases like std::string.
> 
> Switching to std::string, for fmt::join should work.

good point. why did you decide to not implement a similar growth factor in sstring? I wonder if there are other places that might be affected by this too.

doesn't fmt::join() lazily evaluate the range elements?

Reply:
@avikivity I don't think the switch to std::string is strictly needed, now that we switched to sstring.

fmt::join() doc says:
"Returns a view that formats `range` with elements separated by `sep`." 

But I will switch to std::string, it doesn't hurt and makes it quadratic behavior proof.

Reply:
Of course, I meant: "now that we switch to fmt::join()"

Reply:
Right,  fmt::join produces a range that, when evaluated by the formatter, formats the elements one at a time onto the output, and the separator in between the elements, so it won't even produce the concatanated temporary string like boost join did.

Reply:
after auditing all the caller sites of `formatted_sstables_list`, i think we can assume that the `vector<shared_sstables>` instances formatted by `formatted_sstables_list` always outlive the latter. so we can go even further by lazily evaluate the return value of `to_string(const shared_sstable&, bool)` by specializing a `fmt::formatter` for `formatted_sstables_list`. please note, in this change, we still materialize all the return value of `to_string()` into a preallocated vector. if we are allowed to define `to_string()` in the header file as part of the formatter, we can even write the formatted string directly into the buffer provided by logger. we can leave this to yet another change though.

now that this change has been queued. i will create a follow-up change to implement this proposal.

Reply:
A followup is good. thanks.

Reply:
> we can even write the formatted string directly into the buffer provided by logger. we can leave this to yet another change though.

In theory, we could ditch `formatted_sstables_list` altogether and just `fmt::format("{}", _sstables)`
where `_sstables` is e.g. `std::vector<sstables::shared_sstable>`, and that would format all sstables directly onto the output stream, since there is a formatter for shared_sstables that I added..

The problem is that this might stall with a large number of sstables, hence we have `formatted_sstables_list` that collects the per-sstable formatted string and allows preemption in the loop to prevent stalls.
Then, formatting the preformatted strings into the output is cheaper as it only requires copying the strings onto the buffer and it is presumed to be stall-safe.

Reply:
i see, so to avoid the potentially large latency in the last `fmt::format("{}", _sstables)` call, we trade the large stallment for the formatting and the deep copies in each iteration. guess this is the optimal solution we can have

Reply:
exactly